### PR TITLE
feat(meshgateway): customize deployments

### LINKED ISSUE #5027: Support customization of the `Deployment`/`Pods` created by `MeshGatewayInstance`
### Description

Similar to #4903, the deployment should be customizable.

### PR DESCRIPTION
**Pull request by johnharris85**
### Checklist prior to review

(Mostly) fixes #5027

- [x] [Link to relevant issue][1] as well as docs and UI issues -- Docs TBC
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) -- Currently not working so skipped (but committed). Need to pair with @michaelbeaumont on it.
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? --
- [x] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch) or add a `ci/` label to run fewer/more tests?

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 1]
```diff
@@ -128,6 +128,11 @@ func (r *GatewayInstanceReconciler) createOrUpdateService(
 				svcAnnotations[k] = v
 			}
 
+			svcLabels := map[string]string{}
+			for k, v := range gatewayInstance.Spec.ServiceTemplate.Metadata.Labels {
+				svcLabels[k] = v
+			}
```
Comment (bartsmykla) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:134]: It looks to me this is unnecessary as I don't see it being used anywhere 🤔

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 2]
```diff
@@ -273,33 +280,75 @@ func (r *GatewayInstanceReconciler) createOrUpdateDeployment(
...
 			}
 
+			if fsGroup := gatewayInstance.Spec.PodTemplate.Spec.PodSecurityContext.FSGroup; fsGroup != 0 {
+				if securityContext == nil {
+					securityContext = &kube_core.PodSecurityContext{
+						FSGroup: &fsGroup,
+					}
+				} else {
+					securityContext.FSGroup = &fsGroup
+				}
```
Comment (bartsmykla) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:290]: Maybe we could simplify it a little bit?
```suggestion
				if securityContext == nil {
					securityContext = &kube_core.PodSecurityContext{}
				}

				securityContext.FSGroup = &fsGroup
```

File: pkg/plugins/resources/k8s/native/api/v1alpha1/mesh_gateway_config.go [hunk 3]
```diff
@@ -86,6 +105,49 @@ type MeshGatewayServiceSpec struct {
...
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// PodSecurityContext corresponds to PodSpec.SecurityContext
+	// +optional
+	PodSecurityContext `json:"securityContext,omitempty"`
+
+	// Container corresponds to PodSpec.Container
+	// +optional
+	Container `json:"container,omitempty"`
```
Comment (michaelbeaumont) [inline pkg/plugins/resources/k8s/native/api/v1alpha1/mesh_gateway_config.go:120]: Are these all supposed to be embedded?

```suggestion
	Container Container `json:"container,omitempty"`
```

Comment (michaelbeaumont) [inline pkg/plugins/resources/k8s/native/api/v1alpha1/mesh_gateway_config.go:120]: Are these all supposed to be embedded?

```suggestion
	Container Container `json:"container,omitempty"`
```

**Review by lukidzi — DISMISSED** [review]

**Review by bartsmykla — APPROVED** [review]

**Comment by lahabana** [comment]
@johnharris85 remember to add docs :)

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 4]
```diff
@@ -273,33 +281,72 @@ func (r *GatewayInstanceReconciler) createOrUpdateDeployment(
...
+			if fsGroup := gatewayInstance.Spec.PodTemplate.Spec.PodSecurityContext.FSGroup; fsGroup != 0 {
+				if securityContext == nil {
+					securityContext = &kube_core.PodSecurityContext{}
+				}
+				securityContext.FSGroup = &fsGroup
+			}
+
+			if readOnlyRFS := gatewayInstance.Spec.PodTemplate.Spec.Container.SecurityContext.ReadOnlyRootFilesystem; readOnlyRFS {
+				container.SecurityContext.ReadOnlyRootFilesystem = &readOnlyRFS
+			}
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:293]:
```suggestion
			container.SecurityContext.ReadOnlyRootFilesystem = &gatewayInstance.Spec.PodTemplate.Spec.Container.SecurityContext.ReadOnlyRootFilesystem
```

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 5]
```diff
@@ -273,33 +281,72 @@ func (r *GatewayInstanceReconciler) createOrUpdateDeployment(
...
+
+			container.VolumeMounts = []kube_core.VolumeMount{
+				{
+					Name:      "tmp-dir",
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:297]: nit but this is its name in the egress/ingress objects

```suggestion
					Name:      "tmp",
```

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 6]
```diff
@@ -273,33 +281,72 @@ func (r *GatewayInstanceReconciler) createOrUpdateDeployment(
...
 				Containers:      []kube_core.Container{container},
 			}
 
+			if serviceAccountName := gatewayInstance.Spec.PodTemplate.Spec.ServiceAccountName; serviceAccountName != "" {
+				podSpec.ServiceAccountName = serviceAccountName
+			}
+
+			podSpec.Volumes = []kube_core.Volume{
+				{
+					Name: "tmp-dir",
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:313]:
```suggestion
					Name: "tmp",
```

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+		}).Build()
+
+		reconciler = &GatewayInstanceReconciler{
+			Client: kubeClient,
+			Log:    core.Log.WithName("test"),
+		}
+	})
+
+	It("should create correct service and deployment objects", func() {
+		Skip("Currently broken") // TODO: Fix this w/@mikebeaumont
```
Comment (michaelbeaumont) [inline pkg/plugins/runtime/k8s/controllers/gateway_instance_controller_test.go:96]: @johnharris85 should we look at it?

Comment (lahabana) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller_test.go]: I've reworked things a little bit to make this testable. PTAL

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go [hunk 8]
```diff
@@ -71,17 +69,41 @@ func (r *GatewayInstanceReconciler) Reconcile(ctx context.Context, req kube_ctrl
 	mesh := k8s_util.MeshOfByLabelOrAnnotation(r.Log, gatewayInstance, &ns)
 
 	orig := gatewayInstance.DeepCopyObject().(kube_client.Object)
-	svc, err := r.createOrUpdateService(ctx, mesh, gatewayInstance)
+	gatewayK8sList := mesh_k8s.MeshGatewayList{}
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:72]: also isn't this completely broken for multi zone? it now only lists MeshGateways in the local cluster

[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:72]: I think we should revert the rewrite and leave out the tests or add them after in a separate PR.

[the remark this answers is not in the document]
Comment (lahabana) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller.go:72]: The store is always only reading from local store. It's just that MeshGateways are synced from global to zone. This reconciler is the only reconciler actually using a resource store.

File: pkg/plugins/runtime/k8s/controllers/gateway_instance_controller_test.go (hunk 7 above)
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/controllers/gateway_instance_controller_test.go:96]: Should we delete this file now?

**Pull request merged**